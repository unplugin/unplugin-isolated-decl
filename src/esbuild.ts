/**
 * This entry file is for esbuild plugin.
 *
 * @module
 */

import { IsolatedDecl } from './index'

/**
 * Esbuild plugin
 *
 * @example
 * ```ts
 * import { build } from 'esbuild'
 *
 * build({
 *   plugins: [require('unplugin-isolated-decl/esbuild')()],
 * })
 * ```
 */
const esbuild = IsolatedDecl.esbuild as typeof IsolatedDecl.esbuild
export default esbuild
export { esbuild as 'module.exports' }
