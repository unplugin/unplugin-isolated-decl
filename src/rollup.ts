/**
 * This entry file is for Rollup plugin.
 *
 * @module
 */

import { IsolatedDecl } from './index.ts'

/**
 * Rollup plugin
 *
 * @example
 * ```ts
 * // rollup.config.js
 * import IsolatedDecl from 'unplugin-isolated-decl/rollup'
 *
 * export default {
 *   plugins: [IsolatedDecl()],
 * }
 * ```
 */
const rollup = IsolatedDecl.rollup as typeof IsolatedDecl.rollup
export default rollup
export { rollup as 'module.exports' }
