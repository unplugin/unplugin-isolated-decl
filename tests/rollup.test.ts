import fs from 'node:fs/promises'
import path from 'node:path'
import { rollup } from 'rollup'
import esbuild from 'rollup-plugin-esbuild'
import { describe, expect, test } from 'vitest'
import UnpluginIsolatedDecl from '../src/rollup'

describe('rollup', () => {
  const TEST_SANDBOX_FOLDER = 'temp/rollup'

  test('generate basic', async () => {
    const input = path.resolve(__dirname, 'fixtures/basic/main.ts')
    const dist = path.resolve(__dirname, `${TEST_SANDBOX_FOLDER}/basic`)

    const bundle = await rollup({
      input,
      plugins: [
        UnpluginIsolatedDecl({
          extraOutdir: 'temp',
        }),
        esbuild(),
      ],
      logLevel: 'silent',
    })

    const result = await bundle.generate({
      dir: dist,
    })

    expect(
      result.output.map((asset) =>
        [
          `// ${asset.fileName.replaceAll('\\', '/')}`,
          asset.type === 'chunk' ? asset.code : asset.source,
        ].join('\n'),
      ),
    ).toMatchSnapshot()
  })

  test('write entry-points', async () => {
    const input = {
      a: path.resolve(__dirname, 'fixtures/entry-points/a/index.ts'),
      b: path.resolve(__dirname, 'fixtures/entry-points/b/index.ts'),
    }
    const dist = path.resolve(__dirname, `${TEST_SANDBOX_FOLDER}/entry-points`)

    const bundle = await rollup({
      input,
      plugins: [UnpluginIsolatedDecl(), esbuild()],
      logLevel: 'silent',
    })

    // https://github.com/unplugin/unplugin-isolated-decl/issues/22
    await bundle.write({
      dir: dist,
      entryFileNames: '[name]/index.js',
      preserveModules: true,
    })

    /**
     * Map written output from file system rather than from bundle due to
     * module execution order not consistent
     *
     * @see https://github.com/rollup/rollup/issues/3888
     */
    const allBundledFiles = (
      await fs.readdir(dist, {
        recursive: true,
        withFileTypes: true,
      })
    ).filter((it) => it.isFile())

    const fileSystemOutput = allBundledFiles.map((it) => {
      return (async () => {
        const filePath = path.relative(dist, path.join(it.parentPath, it.name))

        const content = await fs.readFile(path.join(dist, filePath), 'utf-8')

        return [
          `// ${filePath.replaceAll('\\', '/')}`,
          content.toString(),
        ].join('\n')
      })()
    })

    expect(Promise.all(fileSystemOutput)).resolves.toMatchSnapshot()
  })
})
