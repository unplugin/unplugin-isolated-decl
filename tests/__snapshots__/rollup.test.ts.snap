// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rollup 1`] = `
[
  "// main.js
function Component() {
  return /* @__PURE__ */ React.createElement("div", null, "I'm a div in a tsx component!");
}

function hello(s) {
  return "hello" + s;
}
let c = Component;
let num = 1;

export { c, hello, num };
",
  "// temp/main.d.ts
import { type Num } from './types';
export type Str = string;
export declare function hello(s: Str): Str;
export declare let c: React.JSX.Element;
export declare let num: Num;
",
  "// temp/types.d.ts
export type Num = number;
",
  "// temp/types2.d.ts
export type Num2 = number;
",
  "// temp/component.d.ts
export declare function Component(): React.JSX.Element;
",
]
`;
